File: questions/3Sum/3Sum.cpp
// Main file of the Project
#include <bits/stdc++.h>
using namespace std;

void PrintVoV(vector<vector<int>> vec)
{
    // printing the given vector of vector
    for (auto v : vec)
    {
        std::cout << "[ ";
        for (auto &n : v)
        {
            std::cout << n << ", ";
        }
        std::cout << "]" << std::endl;
    }
}

class Solution
{
public:
    vector<vector<int>> threeSum(vector<int> &nums)
    {
        vector<vector<int>> ans;        // creating contianer to hold the answer
        sort(nums.begin(), nums.end()); // sorting the given vector "nums" #O(nlogn)

        for (size_t i = 0; i < nums.size(); i++) // iterating throught the value of given vector #O(n^2)
        {
            if (i && (nums[i] == nums[i - 1])) // if its not the very first element
                continue;                      // if the previous value was equal to the current, skip iteration

            int l = i + 1, r = nums.size() - 1; // creating left and right pointers, for two pointer techinque
            while (l < r)                       // iterate till left is less than right
            {
                int threeSum = nums[i] + nums[l] + nums[r]; // storing the sum of three values
                if (threeSum > 0)                           // if sum is greater than 0 decrementing right pointer
                    r--;
                else if (threeSum < 0) // if sum is less then 0 increment left pointer
                    l++;
                else // if sum is equals to 0
                {
                    vector<int> v = {nums[i], nums[l], nums[r]};
                    ans.push_back(v); // pushing the triple vector in "ans" vector
                    l++;              // incrementing the left pointer till it points any new value
                                      // ( skipping duplicates in following loop)
                    while ((nums[l] == nums[l - 1]) && (l < r))
                        l++;
                }
            }
        }
        PrintVoV(ans);
        // return the "ans" vector of vector
        return ans;
    }
};

int main(int arg, char *argv[])
{
    Solution s;
    vector<int> v = {-1, 0, 1, 2, -1, -4, -2, -3, 3, 0, 4};
    s.threeSum(v);
    return 0;
}

File: questions/3Sum/3Sum.js
function threeSum(nums) {
    //   console.log("Given length is : ", nums.length);
    var ans = []; // this will hold the answer array
    //   nums.sort(); // sorting the given array, this default sort is useless never use it!!
    nums.sort(function (a, b) { return a - b; });
    nums.forEach(function (e, i) {
        // if array element is not same as previous one
        if (!(i && nums[i] === nums[i - 1])) {
            var left = i + 1; // left pointer and right pointer to apply two pointer technique
            var right = nums.length - 1;
            // iterate till left is less than right pointer
            while (left < right) {
                var threeSum_1 = nums[i] + nums[left] + nums[right]; // sum of three numbers
                if (threeSum_1 > 0)
                    right -= 1; // is sum is greater than 0, decrement right
                else if (threeSum_1 < 0)
                    left += 1; // if sum is less than 0, increment left
                else {
                    // else pushing the triplet in the a new array
                    // let n = [nums[i], nums[left], nums[right]];
                    ans.push([nums[i], nums[left], nums[right]]); // pushing new array to answer array
                    left += 1; // incrementing the left pointer till points new unique value
                    while (nums[left] === nums[left - 1] && left < right)
                        left += 1;
                }
            }
        }
    });
    // returning the answer array
    return ans;
}
console.log(threeSum([-1, 0, 1, 2, -1, -4, -2, -3, 3, 0, 4]));
console.log(threeSum([-1, 0, 1, 2, -1, -4]));
console.log(threeSum([-1, -1, -4, 0, 1, 2]));
console.log(threeSum([0, 0]));
